src/docfx/lib/JsonUtility.cs

```diff
@@ -125,6 +121,103 @@ public static JObject Merge(JObject first, IEnumerable<JObject> objs)
+        {
+            var errors = new List<Error>();
+            var nullNodes = new List<JToken>();
+            token.Traverse(errors, mappings, nullNodes);
+            foreach (var node in nullNodes)
+            {
+                node.Remove();
+            }
+            return (errors, token);
+        }
+
+        private static bool IsNullOrUndefined(this JToken token)
+        {
+            return
+                (token == null) ||
+                (token.Type == JTokenType.Null) ||
+                (token.Type == JTokenType.Undefined);
+        }
+
+        /// <summary>
+        /// Parse a string to JToken.
+        /// Validate null value during the process.
+        /// </summary>
+        private static (List<Error>, JToken) Deserialize(TextReader reader)
+        {
+            using (JsonReader json = new JsonTextReader(reader))
+            {
+                return DefaultDeserializer.Deserialize<JToken>(json).ValidateNullValue();
+            }
+        }
+
+        private static void Traverse(this JToken token, List<Error> errors, JTokenSourceMap mappings, List<JToken> nullNodes)
+        {
+            if (token is JArray array)
+            {
+                TraverseArray(token, errors, mappings, nullNodes, array);
+            }
+            else
+            {
+                TraverseProperty(token, errors, mappings, nullNodes);
+            }
+        }
+
+        private static void TraverseArray(JToken token, List<Error> errors, JTokenSourceMap mappings, List<JToken> nullNodes, JArray array)
+        {
+            foreach (var item in token.Children())
+            {
+                if (item.IsNullOrUndefined())
+                {
+                    LogWarningForNullValue(array, errors, mappings);
+                    nullNodes.Add(item);
+                }
+                else if (item is JArray)
+                {
+                    item.Traverse(errors, mappings, nullNodes);
+                }
+                else
+                {
+                    TraverseProperty(item, errors, mappings, nullNodes);
+                }
+            }
+        }
+
+        private static void TraverseProperty(JToken token, List<Error> errors, JTokenSourceMap mappings, List<JToken> nullNodes)
+        {
+            foreach (var item in token.Children())
+            {
+                var prop = item as JProperty;
+                if (prop.Value.IsNullOrUndefined())
+                {
+                    LogWarningForNullValue(token, errors, mappings);
+                    nullNodes.Add(item);
+                }
+                else if (prop.Value is JArray)
+                {
+                    prop.Value.Traverse(errors, mappings, nullNodes);
+                }
+            }
+        }
+
+        private static void LogWarningForNullValue(JToken item, List<Error> errors, JTokenSourceMap mappings)
+        {
+            if (mappings == null)
```

Comment:
This should never be null ?

Reply:
This is not null for Yaml parsing, null for Json parsing

---
In reply to: [200284005](https://github.com/dotnet/docfx/pull/2994#discussion_r200284005) [](ancestors = 200284005)